orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -798,6 +800,156 @@ class PipelineSplitter {
   }
 };
 
+// TODO: move to a proper location for pipeline training
+struct PipelineBatchInfo {
+  std::vector<std::pair<int64_t, int64_t>> events;
+  int64_t retired_batch;  // index of retired batch, so its data could be reused
+};
+
+class PipelineBatchPlanner {
+ private:
+  int64_t max_id_;
+
+ public:
+  PipelineBatchPlanner()
+      : max_id_(::onnxruntime::contrib::OrtEventPool::GetPoolSize() - 1) {
+  }
+
+  void CreatePlan(const std::vector<int64_t>& start_ids,
+                  size_t num_batches,
+                  std::vector<PipelineBatchInfo>& plan) {
+    size_t num_stages = start_ids.size();
+
+    // generate timeline
+    Timeline timeline = GenerateTimeline(num_stages, num_batches);
+
+    // fill in plan
+    plan.resize(num_batches);
+    for (size_t batch = 0; batch < num_batches; ++batch) {
+      auto& info = plan[batch];
+      info.events.resize(2 * num_stages - 1);  // two pairs for each stage except for last stage, which only need one
+      info.retired_batch = -1;
+    }
+
+    // each stage requires 2 pair of wait/record events for FW/BW
```

Comment:
Just rewording...
```suggestion
Each batch-stage computes `Wait` for the previous stage's `FW` --> `FW` --> `Record` to signal the next batch-stage's `BW` --> `Wait` for the next batch-stage's `BW` --> `BW` --> `Record` to signal the `BW` of the previous stage. Thus, each stage requires 2 wait-record pairs. The last stage only needs one pair because it has no downstream stage.```

Reply:
Wait/Record here is only for compute resource exclusiveness, it has nothing to do with data dependency across pipeline stages. Data dependency is maintained separately. and in parallel with pipeline execution to hide latency.

---
In reply to: [395763399](https://github.com/microsoft/onnxruntime/pull/3206#discussion_r395763399) [](ancestors = 395763399)